modules/top-level/plugins/combine-plugins.nix

```
@@ -0,0 +1,102 @@
+{ lib, pkgs, ... }:
+let
+  inherit (import ./utils.nix) normalizePlugin normalizePluginList;
+
+  # Remove dependencies from all plugins in a list
+  __removeDependencies = ps: map (p: p // { plugin = removeAttrs p.plugin [ "dependencies" ]; }) ps;
+
+  # Combine a list of plugins into a single normalized plugin "pack"
+  __mkPluginPack =
+    { pluginsToCombine, pathsToLink }:
+    let
+      # Every plugin has its own generated help tags (doc/tags)
+      # Remove them to avoid collisions, new help tags
+      # will be generate for the entire pack later on
+      overriddenPlugins = map (
+        plugin:
+        plugin.plugin.overrideAttrs (prev: {
+          nativeBuildInputs = lib.remove pkgs.vimUtils.vimGenDocHook prev.nativeBuildInputs or [ ];
+          configurePhase = ''
+            ${prev.configurePhase or ""}
+            rm -vf doc/tags'';
+        })
+      ) pluginsToCombine;
+
+      # Gather python 3 dependencies from every plugins
+      python3Dependencies =
+        let
+          deps = map (p: p.plugin.python3Dependencies or (_: [ ])) pluginsToCombine;
+        in
+        ps: builtins.concatMap (f: f ps) deps;
+
+      # Combined plugin
+      combinedPlugin = pkgs.vimUtils.toVimPlugin (
+        pkgs.buildEnv {
+          name = "plugin-pack";
+          paths = overriddenPlugins;
+          inherit pathsToLink;
+          # Remove empty directories and activate vimGenDocHook
+          postBuild = ''
+            find $out -type d -empty -delete
+            runHook preFixup
+          '';
+          passthru = {
+            inherit python3Dependencies;
+          };
+        }
+      );
+
+      # Combined plugin configs
+      combinedConfig = builtins.concatStringsSep "\n" (
+        builtins.concatMap (x: lib.optional (x.config != null && x.config != "") x.config) pluginsToCombine
+      );
+    in
+    normalizePlugin {
+      plugin = combinedPlugin;
+      config = combinedConfig;
+    };
+
+  # Take a list of combined plugins, combine the relevant ones and return the resulting list of plugins
+  combinePlugins =
+    {
+      standalonePlugins,
+      pathsToLink,
+      normalizedPlugins,
+    }:
+    let
+      # Plugin list extended with dependencies
+      allPlugins =
+        let
+          pluginWithItsDeps =
+            p: [ p ] ++ builtins.concatMap pluginWithItsDeps (normalizePluginList p.plugin.dependencies or [ ]);
```

The one time where I like to have uneccessary parentheses is when passing `or` to a function.

TLDR; `or` is one of the only nix operators that has a higher priority than function application. Since you need parentheses for almost all other cases of applying an expression to a function, I think they aid readability when used with `or` too.

```suggestion
            p: [ p ] ++ builtins.concatMap pluginWithItsDeps (normalizePluginList (p.plugin.dependencies or [ ]));
```

Alternatively, it might be nice to have a separate "get dependencies" function:

```nix
getDeps = p: p.plugin.dependencies or [ ];
```